### PR TITLE
[linked-list] range-based `for` loop iteration over all entries

### PR DESCRIPTION
**Pull request by abtink**
This commit enables use of range-based `for` loop for iterating over
all the entries in a `LinkedList`. It adds private implementations of
`Iterator` and `ConstIterator` which inherit from `ItemPtrIterator`
and use the `GetNext()` to go to the next entry. The range-based
`for` loop practically acts as a syntactic sugar helping simplify the
code and does not add any code size (or any expected run-time)
overhead.

### CODE REVIEW

**Review by simonlingoogle — APPROVED** [review]
LGTM 👍

File: src/core/net/srp_server.cpp [hunk 1]
```diff
@@ -855,15 +855,15 @@ Error Server::ProcessServiceDescriptionInstructions(Host &                   aHo
     // that `mTimeLastUpdate` on a new `Service::Description` is set to
     // `GetNow().GetDistantPast()`.
 
-    for (desc = aHost.mServiceDescriptions.GetHead(); desc != nullptr; desc = desc->GetNext())
+    for (Service::Description &dsc : aHost.mServiceDescriptions)
```
Comment (wgtdkp) [inline src/core/net/srp_server.cpp:858]: nit:
```suggestion
    for (Service::Description &desc : aHost.mServiceDescriptions)
```

**Review by wgtdkp — APPROVED** [review]
Looks really nicer! 👍

File: src/core/net/srp_server.cpp (hunk 1 above)
Comment (abtink) [reply src/core/net/srp_server.cpp:858]: We already have `desc` variable in the scope above in this method, so if we re-use the same name in the loop, we get a warning/error that the loop variable is shadowing an existing variable name.

```
src/core/net/srp_server.cpp:858:32: error: declaration shadows a local variable [-Werror,-Wshadow]
    for (Service::Description &desc : aHost.mServiceDescriptions)
```

Comment (wgtdkp) [reply src/core/net/srp_server.cpp:858]: oh, I missed this fact. But it looks more like a typo to name it `dsc` :(

What about moving the `Service::Description *desc;` into the `for` loop above? As that `desc` is never referenced out of that loop.

Comment (abtink) [reply src/core/net/srp_server.cpp:858]: Yes. I think that can work. Good idea. Will change in next push.

Comment (wgtdkp) [reply src/core/net/srp_server.cpp]: Thanks!

**Review by jwhui — APPROVED** [review]

**Pull request merged**